### DISCUSSION TITLE
Proper way to incorporate transforms into scenes when using cuda

### DISCUSSION THREAD

**Discussion by ShnitzelKiller** [mitsuba-renderer/mitsuba3#622] [Q&A] [answered]
I notice that in the tutorials, it is expected that we explicitly import `from mitsuba.scalar_rgb import Transform4f as T` even if we are not using the scalar_rgb variant. The function `mi.load_dict()` throws an error if the scene dict contains instances of transforms from the cuda variant, even if that is what you are using to render the scene. Why is this? It seems like a weird hack/workaround, and my IDE hates it since the python linter cannot resolve the name mitsuba.scalar_rgb.  Is there a better way to do things?

This works for 3D transforms (4x4 matrices), but I feel like this might somehow be the wrong approach, since I cannot follow a basic example on transforming a texture using the 'to_uv' attribute, which accepts a ScalarTransform3f, or so it seems. When I provide an object of the type mi.scalar_rgb.ScalarTransform3f, I get the error
`[xml_v.cpp:334] Unkown value type: <class 'mitsuba.scalar_rgb.ChainTransform3f'>`.

**Comment by njroussel** [comment, marked as answer]
1) The scene parsing is all done without the JIT and on the CPU. It's a scalar CPU process. Having it be able to deal with JIT types would make this piece of code variant-dependent. In addition, specifically for cuda variants, an implicit and costly GPU->CPU memory transfer would take place.
The current solution, although verbose, makes memory migrations explicit and keeps the scene loading variant independent.

2) I don't know if your linter is being aggressive, but the `mitsuba/stubs/` folder does have information for `scalar_rgb` in the `scalar_rgb.pyi` file. If this is truly annoying I'd welcome any information about why this is currently not working in your setup.

3) I litterally just noticed this now, the scene loader can only handle `Transform4f` variables. The proper matrix size is then extracted in the plugin constructors. It's less obvious in the XML representation. Basically, you should always pass `mi.scalar_rgb.ScalarTransform4f` objects even when only 3 dimensions are used. I will fix the documentation for this right away.
